fairscale/optim/adascale.py

```diff
@@ -322,33 +428,38 @@ def _final_callback(self) -> None:
 
         # Compute the sums of squares for reduced gradients.
         # Divide by _num_grads_to_accum since the gradients are accumulated.
-        #
-        # Note: we are mutating the gradients here!!!
         total_grad_sqr = np.array(
-            [
-                sum(param.grad.div_(self._num_grads_to_accum).pow(2).sum().item() for param in group["params"])
-                for group in self._optimizer.param_groups
-            ]
+            [sum(param.grad.pow(2).sum().item() for param in group["params"]) for group in self._optimizer.param_groups]
         )
+        # Divide by (_num_grads_to_accum ** 2) to account for gradient
+        # accumulation.
+        if self._num_grads_to_accum > 1:
+            # np array doesn't support /=.
+            total_grad_sqr = total_grad_sqr / (self._num_grads_to_accum ** 2)
 
         # Wait for all_reduce to be done and move it to cpu & np.
         if work:
             work.wait()
         local_grad_sqr = self._local_grad_sqr.cpu().numpy()
-        self._local_grad_sqr = None
 
         # See appendix B.3 of the paper.
+        # Modified to handle cases where scale != world_size
         #
-        # local_grad_sqr is \sigma_{i=1}^{S}\norm{g_t_i}
-        # total_grad_sqr is \norm{g_t}
+        # local_grad_sqr is \sum_{i=1}^{c N} \norm{g_t_i}^2
+        # where N is world size and c is num_grads_to_accum
+        # total_grad_sqr is \norm{\bar{g}_t}^2
         S = self._scale
-        grad_var = local_grad_sqr / (S - 1) - total_grad_sqr * S / (S - 1)
+        cN = self._world_size * self._num_grads_to_accum
+        grad_var = local_grad_sqr * (S / cN) / (cN - 1) - total_grad_sqr * S / (cN - 1)
         grad_sqr = total_grad_sqr - grad_var / S
         grad_var = np.maximum(grad_var, 1e-6)
         grad_sqr = np.maximum(grad_sqr, 0.0)
-        theta = self._smoothing
+        theta = self.smoothing
```

Comment:
is `theta` really needed here ?

Reply:
Good point! Will simplify it. Thanks for the speedy review! I am not in a hurry to merge. Will wait for Mike. :-)